campaignion_layout/src/Themes.php

```diff
@@ -57,18 +57,44 @@ public function getTheme(string $theme_name = NULL) {
   }
 
   /**
-   * Get all themes that are enabled and have the feature enabled as well.
+   * Get all themes as theme objects.
    */
-  public function enabledThemes() {
+  public function allThemes() {
     $self = $this;
-    $all_themes = array_map(function ($theme) use ($self) {
+    return array_map(function ($theme) use ($self) {
       return $self->getTheme($theme->name);
     }, $this->themes);
-    return array_filter($all_themes, function ($theme) {
+  }
+
+  /**
+   * Get all themes that are enabled and have the feature enabled as well.
+   */
+  public function enabledThemes() {
+    return array_filter($this->allThemes(), function ($theme) {
       return $theme->isEnabled() && $theme->hasFeatureEnabled();
     });
   }
 
+  /**
+   * Get all themes which’s hooks should be invoked.
+   *
+   * The hook should be invoked on all enabled themes and all their base themes.
+   * Even disabled base themes should get their hook invoked as this matches how
+   * Drupal core treats themes.
+   */
+  protected function hookThemes() {
+    $enabled_themes = array_filter($this->allThemes(), function ($theme) {
+      return $theme->isEnabled() && $theme->hasFeature();
+    });
+    $hook_themes = $enabled_themes;
```

Comment:
This could just use the previous function: `$hook_themes = $this->enabledThemes()`?

Reply:
Not quite. `$enabled_themes` doesn’t (and shouldn’t) check whether the feature toggle is enabled in the settings. So this is not the same as `$this->enabledThemes()`.

Reply:
What about renaming the variable to `$layout_themes`?

Reply:
Ah, yes, `hasFeature()` vs. `hasFeatureEnabled()`, that’s easy to overlook…

I like that the variable matches the function name, but `layoutThemes()` could be better in providing information what those themes are used *for* instead of *where*?

Reply:
Another thought to make this simpler. The main complication of working out the list of themes upfront is caused by the assumption that each theme should be invoked only once (mainly because `drupal_array_merge_deep($array, $array) == $array` doesn’t hold for all inputs). We could instead do the following:

1. `hook_campaignion_layout_info()` gets the base theme’s return value (or an empty array) as parameter.
2. The themes are responsible for keeping or discarding their base theme’s layouts.
3. If a theme doesn’t implement the hook then the base theme’s return value is returned.

Disadvantage: This would mean layouts could appear/disappear based on whether the base theme is enabled. ^^

Reply:
I like the parameter idea, but we'd need to call the hook of disabled base themes…which probably isn’t possible?

Reply:
Surely possible … and exactly what’s happening with `Themes::hookThemes()` right now. The difference between `$enabled_themes` and `$hook_themes` is exactly the disabled base themes.